### PR TITLE
[TE] Disable alerts if it has no success run within 30 days

### PR DESCRIPTION
**Pull request by xiaohui-sun**
Check alert's lastTaskRunTime. If there is no success run within 30 days then disable it.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java [hunk 1]
```diff
@@ -120,11 +134,62 @@ private void executeMonitorUpdate(MonitorTaskInfo monitorTaskInfo) {
...
+
     } catch (Exception e) {
       LOG.error("Exception in monitor update task", e);
     }
   }
 
+  /**
+   * Disable the alert if it was updated before {@MAX_TASK_FAIL_DAYS} but there is no success run since then.
+   */
+  private void disableLongFailedAlerts() {
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java:149]: Can we include some unit tests for this?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java [hunk 2]
```diff
@@ -120,11 +134,62 @@ private void executeMonitorUpdate(MonitorTaskInfo monitorTaskInfo) {
...
+    DetectionConfigManager detectionDAO = DAO_REGISTRY.getDetectionConfigManager();
+    List<DetectionConfigDTO> detectionConfigs = detectionDAO.findAllActive();
+    long currentTimeMills = System.currentTimeMillis();
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java:152]: nit: typo `currentTimeMillis` & `maxTaskFailMillis`

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java [hunk 3]
```diff
@@ -120,11 +134,62 @@ private void executeMonitorUpdate(MonitorTaskInfo monitorTaskInfo) {
...
+    long maxTaskFailMills = TimeUnit.DAYS.toMillis(MAX_FAILED_DISABLE_DAYS);
+    for (DetectionConfigDTO config : detectionConfigs) {
+      try {
+        Timestamp updateTime = config.getUpdateTime();
+        if (updateTime != null && config.getHealth() != null && config.getHealth().getDetectionTaskStatus() != null) {
+          long lastTaskExecutionTime = config.getHealth().getDetectionTaskStatus().getLastTaskExecutionTime();
+          if (updateTime.getTime() <= currentTimeMills - maxTaskFailMills && (lastTaskExecutionTime == -1L
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java:159]: Can we skip the `lastTaskExecutionTime == -1L` check?
Isn't that already covered by `lastTaskExecutionTime <= currentTimeMills - maxTaskFailMills`?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java [hunk 4]
```diff
@@ -120,11 +134,62 @@ private void executeMonitorUpdate(MonitorTaskInfo monitorTaskInfo) {
...
+              || lastTaskExecutionTime <= currentTimeMills - maxTaskFailMills)) {
+            config.setActive(false);
+            detectionDAO.update(config);
+            sendDisableAlertNotificationEmail(config);
+            LOG.info("Disabled alert " + config.getId() + " since it failed more than " + MAX_FAILED_DISABLE_DAYS + " days");
+            LOG.info("Task last update time: " + config.getUpdateTime());
+            LOG.info("Last success task execution time: " + lastTaskExecutionTime);
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java:166]: Clubbing them into one log statement might be easier to debug.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java [hunk 5]
```diff
@@ -120,11 +134,62 @@ private void executeMonitorUpdate(MonitorTaskInfo monitorTaskInfo) {
...
+      // disable alerts that failed consecutively for a long time
+      disableLongFailedAlerts();
+
     } catch (Exception e) {
       LOG.error("Exception in monitor update task", e);
     }
   }
 
+  /**
+   * Disable the alert if it was updated before {@MAX_TASK_FAIL_DAYS} but there is no success run since then.
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java:147]: What if the alert was updated(after receiving the below notification) and it still continues to fail? We don't want to disable & notify them again?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java (hunk 3 above)
Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java:159]: lastTaskExecutionTime == -1L is used for backward compatibility.
Currently we have many long failing alerts have -1L.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java (hunk 5 above)
Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java:147]: Yes we want to handle it more conservatively. Give them 30 days grace period.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java (hunk 2 above)
Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java]: fixed

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java [hunk 6]
```diff
@@ -120,11 +134,62 @@ private void executeMonitorUpdate(MonitorTaskInfo monitorTaskInfo) {
...
+            config.setActive(false);
+            detectionDAO.update(config);
+            sendDisableAlertNotificationEmail(config);
+            LOG.info("Disabled alert " + config.getId() + " since it failed more than " + MAX_FAILED_DISABLE_DAYS + " days");
+            LOG.info("Task last update time: " + config.getUpdateTime());
+            LOG.info("Last success task execution time: " + lastTaskExecutionTime);
+          }
+        }
+      } catch (Exception e) {
+        LOG.error("Exception in disabling alert ", config.getId());
```
[the remark this answers is not in the document]
Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java]: fixed

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java (hunk 1 above)
Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/monitor/MonitorTaskRunner.java:149]: Currently we don't have a good way to test it. The only way would be integration test. But having more than one integration tests running may cause conflicts. That's why I have to add sequence number in our local integration test.

**Pull request merged**